external-crates/move/crates/move-compiler/src/hlir/match_compilation.rs

```diff
@@ -0,0 +1,1903 @@
+        vec![Some(ty.clone())],
+        Box::new(make_copy_exp(ty, rhs.var.loc, rhs.var)),
+    );
+    sp(lhs_loc, binder)
+}
+
+fn make_lit_test(lit_exp: T::Exp, value: Value) -> T::Exp {
+    let loc = value.loc;
+    let value_exp = Box::new(T::exp(
+        lit_exp.ty.clone(),
+        sp(loc, T::UnannotatedExp_::Value(value)),
+    ));
+    let bool = N::Type_::bool(loc);
+    let equality_exp_ = T::UnannotatedExp_::BinopExp(
+        Box::new(lit_exp),
+        sp(loc, BinOp_::Eq),
+        Box::new(bool.clone()),
+        value_exp,
+    );
+    T::exp(bool, sp(loc, equality_exp_))
+}
+
+fn make_if_else(test: T::Exp, conseq: T::Exp, alt: T::Exp, result_ty: Type) -> T::Exp {
+    // FIXME: this span is woefully wrong
+    let loc = conseq.exp.loc;
+    T::exp(
+        result_ty,
+        sp(
+            loc,
+            T::UnannotatedExp_::IfElse(Box::new(test), Box::new(conseq), Box::new(alt)),
+        ),
+    )
+}
+
+fn make_copy_exp(ty: Type, loc: Loc, var: Var) -> T::Exp {
+    let exp_ = T::UnannotatedExp_::Copy {
+        var,
+        from_user: false,
+    };
+    T::exp(ty, sp(loc, exp_))
+}
+
+fn make_freeze_exp(ty: Type, loc: Loc, arg: T::Exp) -> T::Exp {
+    let freeze_fn = Box::new(sp(loc, T::BuiltinFunction_::Freeze(ty.clone())));
+    let freeze_exp = T::UnannotatedExp_::Builtin(freeze_fn, Box::new(arg));
+    T::exp(ty, sp(loc, freeze_exp))
+}
+
+fn make_deref_exp(ty: Type, loc: Loc, arg: T::Exp) -> T::Exp {
+    let deref_exp = T::UnannotatedExp_::Dereference(Box::new(arg));
+    T::exp(ty, sp(loc, deref_exp))
+}
+
+//------------------------------------------------
+// Counterexample Generation
+//------------------------------------------------
+
+#[derive(Clone, Debug)]
+enum CounterExample {
+    Wildcard,
+    Literal(String),
+    Constructor(DatatypeName, VariantName, Vec<CounterExample>),
+    Note(String, Box<CounterExample>),
+}
+
+impl CounterExample {
+    fn into_notes(self) -> VecDeque<String> {
+        match self {
+            CounterExample::Wildcard => VecDeque::new(),
+            CounterExample::Literal(_) => VecDeque::new(),
+            CounterExample::Note(s, next) => {
+                let mut notes = next.into_notes();
+                notes.push_front(s.clone());
+                notes
+            }
+            CounterExample::Constructor(_, _, inner) => inner
+                .into_iter()
+                .flat_map(|ce| ce.into_notes())
+                .collect::<VecDeque<_>>(),
+        }
+    }
+}
+
+impl Display for CounterExample {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CounterExample::Wildcard => write!(f, "_"),
+            CounterExample::Literal(s) => write!(f, "{}", s),
+            CounterExample::Note(_, inner) => inner.fmt(f),
+            CounterExample::Constructor(e, v, args) => {
+                write!(f, "{}::{}", e, v)?;
+                if !args.is_empty() {
+                    write!(f, "(")?;
```

Comment:
Not sure about this. I think we may need to keep positional vs. named (as a flag) for each variant, and use that info here when reconstructing the counterexample so that we output it in the declaration format since that will match the pattern format. Otherwise folks are going to find the counterexamples confusing I worry.

Reply:
This was not pleasant, but it's done. :)